### DISCUSSION TITLE
How to create a VK_DESCRIPTOR_TYPE_STORAGE_TEXEL_BUFFER

### DISCUSSION THREAD

**Discussion by rolandhill** [vsg-dev/VulkanSceneGraph#841] [Q&A]
Hi All,

I'm trying to create a buffer to store alpha and depth data from a fragment shader so that it can be composited later on in a full screen quad shader. I'm basically trying to implement one of the order independent transparency methods from here https://github.com/nvpro-samples/vk_order_independent_transparency.

To do this I need an A buffer. I've tried to implement it, but can't seam to get it right and it seg faults within the Vulkan SDK. The buffer is created with

```
vsg::ref_ptr<vsg::BufferInfo> createABuffer(vsg::ref_ptr<vsg::Device> device, uint32_t width, uint32_t height)
{
    VkDeviceSize bufferSize = width * height * 8 * sizeof(vsg::uivec4);
    auto buffer = vsg::createBufferAndMemory(device, bufferSize, VK_BUFFER_USAGE_STORAGE_TEXEL_BUFFER_BIT, VK_SHARING_MODE_EXCLUSIVE, VK_MEMORY_PROPERTY_DEVICE_LOCAL_BIT);
    
    auto bufferInfo = vsg::BufferInfo::create();
    bufferInfo->buffer = buffer;
    bufferInfo->offset = 0;
    bufferInfo->range = bufferSize;

    return bufferInfo;
}

```
I then add the descriptors with

```
    descriptorBindings.push_back(VkDescriptorSetLayoutBinding{20, VK_DESCRIPTOR_TYPE_STORAGE_TEXEL_BUFFER, 1, VK_SHADER_STAGE_FRAGMENT_BIT, nullptr}); // A Buffer

    auto TexelBuffer = vsg::DescriptorBuffer::create(vsg::BufferInfoList{aBuffer}, 20, 0, VK_DESCRIPTOR_TYPE_STORAGE_TEXEL_BUFFER);
    descriptors.push_back(TexelBuffer);

```

A modified vsgdraw is attached. It compiles and runs then seg faults.

I've been through all the examples and can't figure out where I'm going wrong. Any advice is appreciated. 

Thanks and regards,
Roland
[vsgdraw.zip](https://github.com/vsg-dev/VulkanSceneGraph/files/11667335/vsgdraw.zip)

**Comment by robertosfield** [comment]
I can't look at it right away so just a quick reply for now.  Try enabling the Vulkan debug layer to see if anything helpful is picked up.

**Comment by rolandhill** [comment]
Hi Robert,

No problem. I appreciate the help when you have time. I do have the Vulkan
debug layer enabled and there is nothing reported.

Roland

**Comment by rolandhill** [comment]
I figured out the seg fault problem. I was trying to use a vsg::BufferInfo when I should have been using a vsg::BufferView. 

Now to try and make it do something useful! :)

**Reply by robertosfield** [reply]
Good to hear you figured it out.
